### PR TITLE
Provisioning / entitlements refactor

### PR DESCRIPTION
**Pull request by neilk**
The main difference is that we have gone from 1 provisioning profile/entitlements to potentially many. This is required to support sub-bundles with different bundle ids + distribution provisioning profiles, such as WatchKit.

A dictionary of provisioning profiles is kept in Provisioner, as well as entitlements as strings. Provisioner is then passed down recursively to all sub-bundles, and each bundle will get the best provisioning profiles and entitlements for itself.

Entitlements have also been refactored to be a little neater; there is no more Entitlements.plist file that gets added to the app. This was always a wart of isign & I'm glad it's gone :)

A few other cleanups such as moving OpenSslShell to its own class.

### CODE REVIEW

File: isign/openssl_shell.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
+""" This is for a few random commands we either can't or haven't figured out how to do
+    with crypto libraries """
```
Comment (sah) [inline isign/openssl_shell.py]: Good call isolating this stuff.

File: docs/submit.md [hunk 2]
```diff
@@ -279,3 +280,95 @@ TBD: add docs for such properties here.
...
+	</array>
+	
+?? is this what is selected for other apps??
+
+The benefit of using a wildcard cert is that we wouldn't have to worry about creating a cert for each
+individual app
+On the other hand, what is the normal practice for Square??
+...We could check in the actual Square app that we have
+Actual Square Apps use different entitlements for app, watch-appex, watch-app.
```
Comment (segiddins) [inline docs/submit.md]: are these just notes, or do you mean to mention Square ?

File: isign/bundle.py [hunk 3]
```diff
@@ -198,71 +202,34 @@ class App(Bundle):
...
-        self.entitlements_path = entitlements_path
-
-    def resign(self, deep, cms_signer, provisioning_profile, alternate_entitlements_path=None):
-        """ signs app in place """
 
-        # TODO all this mucking about with entitlements feels wrong. The entitlements_path is
-        # not actually functional, it's just a way of passing it to later stages of signing.
-        # Maybe we should determine entitlements data in isign/archive.py or even isign/isign.py,
-        # and then embed it into CmsSigner?
+    def provision(self, team_id, provisioner):
```
Comment (sah) [inline isign/bundle.py]: I don't have a specific recommendation here, but when I see OO code that reads as a reflexive / redundant statement like "Bundle, provision yourself using this Provisioner", I always wonder if there's a clearer structure. I.e., why do Bundles know how to provision when there is a thing called a Provisioner? 

I dunno, maybe consider whether the method belongs on Provisioner, or whether Provisioner should be called something else if it's not the home for this method. I don't feel strongly about it.

File: docs/submit.md [hunk 4]
```diff
@@ -279,3 +280,95 @@ TBD: add docs for such properties here.
...
+And they use fully qualified application identifiers, keychain groups, etc.
+(Square also leaves rando entitlement template files lying around in their app lol)
```
Comment (segiddins) [inline docs/submit.md]: it was a mistake! We cleaned it up eventually ;)

File: isign/identifier_matcher.py [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+            if star_index != -1 and star_index != len(pattern_parts) - 1:
+                raise BadIdentifier("pattern has a non-terminal asterisk: {}".format(pattern))
+        except ValueError:
+            pass
+
+        # to be a match, there must be equal or fewer pattern parts. Neither 'foo.bar' nor 'foo.*' can match 'foo'
+        i = 0
+        p = 0
+        while i < len(identifier_parts) or p < len(pattern_parts):
+            if p < len(pattern_parts) and i < len(identifier_parts):
```
Comment (sah) [inline isign/identifier_matcher.py]: Maybe flip these to match the order in the while condition above.

File: isign/identifier_matcher.py [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+                    i = i + 1
+                    p = p + 1
+                    score = score + 1
+                    continue
+                elif pattern_parts[p] == '*':
+                    break
+            score = 0
+            break
+
+        return score
```
Comment (sah) [inline isign/identifier_matcher.py]: No newline at eof.

File: isign/identifier_matcher.py [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+        p = 0
+        while i < len(identifier_parts) or p < len(pattern_parts):
+            if p < len(pattern_parts) and i < len(identifier_parts):
+                if identifier_parts[i] == pattern_parts[p]:
+                    i = i + 1
+                    p = p + 1
+                    score = score + 1
+                    continue
+                elif pattern_parts[p] == '*':
+                    break
```
Comment (sah) [inline isign/identifier_matcher.py]: I noticed it was possible to break here without ever setting `score`, then had to look quite a few lines above to see that `score` is initialized before the loop. I think it should be initialized just before the loop, with `i` and `p`.

File: isign/identifier_matcher.py [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+            score = 0
+            break
```
Comment (sah) [inline isign/identifier_matcher.py]: Overall, I found the logic around this fall-through condition a little hard to follow. Here's my attempt to clarify this loop, which you can take or leave depending on whether it actually reads clearly to you:

    score = 0
    for identifier_part, pattern_part in itertools.izip_longest(identifier_parts, pattern_parts):
        if identifier_part is None or pattern_part is None:
            score = 0
            break

        if pattern_part == '*':
            break

        if identifier_part != pattern_part:
            score = 0
            break

        score = score + 1

File: isign/provisioner.py [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+        self.app_id_to_pprof = app_id_to_pprof
+        log.debug(app_id_to_pprof)
+        self.app_id_to_entitlements_info = app_id_to_entitlements_info
+        log.debug(self.app_id_to_entitlements_info)
+
+        # each provisioning profile should already have entitlements. But maybe you want to override them.
+        self.app_id_to_entitlements_info.update(self.parse_entitlements(entitlements_paths))
+        # Note, it appears that overriden entitlements paths can only be "less" than the ones in the provisioning
+        # profile. For instance, a property that in the provisioning profile is "foo.*" can be "foo.bar.baz" in the
+        # entitlements, but not the reverse. We could do a sanity check here for that!
```
Comment (sah) [inline isign/provisioner.py]: Could generalize that `IdentifierMatcher` loop I commented on above to check this.

File: tests/test_identifier_matcher.py [hunk 10]
```diff
@@ -0,0 +1,58 @@
+import unittest
+from isign.identifier_matcher import IdentifierMatcher
+from isign.exceptions import BadIdentifier
+
+
+class TestIdentifierMatcher(unittest.TestCase):
```
Comment (sah) [inline tests/test_identifier_matcher.py]: Yesss, unit tests for this. Excellent.

File: isign/identifier_matcher.py (hunk 5 above)
Comment (sah) [reply isign/identifier_matcher.py]: (See below before making changes here.)

File: isign/identifier_matcher.py (hunk 7 above)
Comment (sah) [reply isign/identifier_matcher.py]: (Again, see below before making changes here.)

File: docs/submit.md (hunk 2 above)
Comment (neilk) [reply docs/submit.md]: I inadvertently committed some notes, sorry

File: docs/submit.md (hunk 4 above)
Comment (neilk) [reply docs/submit.md]: For what it's worth, isign has been adding a useless `Entitlements.plist` to every app for 4 years, until this very change.

I'm sorry for the comment, it wasn't very professional. Once I squash and delete this branch it will vanish from history. 

By the way, at Sauce Labs, I would frequently see apps that had similar build scaffolding shipping, right in the app, in one case ballooning its size up to 10%. From very large and prominent companies too.

File: isign/bundle.py (hunk 3 above)
Comment (neilk) [reply isign/bundle.py]: I agree with you and I think we're inching towards inverting control.

There are now so many options with signing that it might be better to have a SigningContext that operates on Bundles/Signables/etc, rather than handing Bundles a set of options and telling them to resign themselves. 

Condensing options into `CmsSigner` and now `Provisioner` has at least reduced the laundry list of things we are passing around.

File: isign/identifier_matcher.py (hunk 8 above)
Comment (neilk) [reply isign/identifier_matcher.py]: This is so much better; amended. Thank you, I was struggling with making that one both easy to read and to pass all tests. And now I know about `izip_longest`; thank you!

File: isign/provisioner.py (hunk 9 above)
Comment (neilk) [reply isign/provisioner.py]: Yup, that's the plan, should is be necessary. Just add a `is_match` which tests for score > 0
